### PR TITLE
Add support to read and write certificates with chain certificates

### PR DESCRIPTION
**Pull request by mrtrev**
This will allow the libstirshaken library to read and write files containing the chain certificates required by some certificates in order to pass validation.  The new functionality has been implemented as such that existing applications will not break and may test for support of these new functions.

These changes are confirmed working in a production environment.

### CODE REVIEW

File: src/stir_shaken_ssl.c [hunk 1]
```diff
@@ -1988,6 +2015,56 @@ X509* stir_shaken_load_x509_from_file(stir_shaken_context_t *ss, const char *nam
...
+		}
+		else {
+			sk_X509_push(cert->xchain, wcert);
+		}
+	}
+
+    if (!cert->x) {
+        snprintf(err_buf, sizeof(err_buf), "No certificate found in file %s", name);
+        stir_shaken_set_error(ss, err_buf, STIR_SHAKEN_ERROR_X509_READ_FROM_FILE);
+        goto fail;
```
[the remark this answers is not in the document]
Comment (mrtrev) [reply src/stir_shaken_ssl.c:2057]: Yes we can do that.  New commit incoming in a moment.

**Comment by andywolk** [comment]
> Unit-tests failed: https://public-artifacts.signalwire.cloud/drone/signalwire/libstirshaken/28/artifacts.html

This PR contains changes from another PR. Please use the branch of the other PR here. Not master.

**Comment by mrtrev** [comment]
Sorry I'm not used to the github workflow.  I'm not sure how to remedy this... I don't think I can change the branch the PR is from.  Should I revert the change and force push this branch or do I need to create a whole new PR from a different branch and close this one?

**Comment by mrtrev** [comment]
Ok I think it is sorted out and back to a coherent PR.

File: src/stir_shaken_ssl.c [hunk 2]
```diff
@@ -1995,6 +2022,60 @@ X509* stir_shaken_load_x509_from_file(stir_shaken_context_t *ss, const char *nam
...
+    char err_buf[STIR_SHAKEN_ERROR_BUF_LEN] = { 0 };
+    FILE *fp = NULL;
+    X509 *wcert = NULL;
+
+    stir_shaken_clear_error(ss);
+
+    if (stir_shaken_zstr(name)) {
+        snprintf(err_buf, sizeof(err_buf), "Cert file name missing");
+        stir_shaken_set_error(ss, err_buf, STIR_SHAKEN_ERROR_X509_CERT_NAME_MISSING_1);
+		goto fail;
```
Comment (andywolk) [inline src/stir_shaken_ssl.c:2036]: Please fix formatting. This is because tabs are mixed with spaces.

File: src/stir_shaken_ssl.c [hunk 3]
```diff
@@ -1768,11 +1768,26 @@ X509* stir_shaken_make_cert_from_public_key(stir_shaken_context_t *ss, EVP_PKEY
...
+}
+
+/**
+ * #define STIR_SHAKEN_CAN_RW_X509_FULLCHAIN is present when stir_shaken_x509_to_disk_fullchain exists
+ */
+
+stir_shaken_status_t stir_shaken_x509_to_disk_fullchain(stir_shaken_context_t *ss, X509 *x, STACK_OF(X509) *xchain, const char *cert_full_name)
 {
     int i = 0;
+    int w = 0;
```
Comment (andywolk) [inline src/stir_shaken_ssl.c:1787]: Declaring a new var could be avoided.

File: src/stir_shaken_ssl.c [hunk 4]
```diff
@@ -1792,12 +1807,24 @@ stir_shaken_status_t stir_shaken_x509_to_disk(stir_shaken_context_t *ss, X509 *x
...
+        if (w == 0) {
             snprintf(err_buf, sizeof(err_buf), "Error writing certificate to file %s", cert_full_name);
             stir_shaken_set_error(ss, err_buf, STIR_SHAKEN_ERROR_FILE_WRITE_2);
             goto fail;
         }
+
+        for (i = 0; i < sk_X509_num(xchain); i++) {
+            xc = sk_X509_value(xchain, i);
+            if (!xc) continue;
+            w = PEM_write_X509(fp, xc);
```
Comment (andywolk) [inline src/stir_shaken_ssl.c:1820]: `if (!PEM_write_X509(fp, xc))` ?

File: src/stir_shaken_ssl.c [hunk 5]
```diff
@@ -1792,12 +1807,24 @@ stir_shaken_status_t stir_shaken_x509_to_disk(stir_shaken_context_t *ss, X509 *x
...
-        if (i == 0) {
+        w = PEM_write_X509(fp, x);
+        if (w == 0) {
             snprintf(err_buf, sizeof(err_buf), "Error writing certificate to file %s", cert_full_name);
             stir_shaken_set_error(ss, err_buf, STIR_SHAKEN_ERROR_FILE_WRITE_2);
             goto fail;
         }
+
+        for (i = 0; i < sk_X509_num(xchain); i++) {
+            xc = sk_X509_value(xchain, i);
```
Comment (andywolk) [inline src/stir_shaken_ssl.c:1818]: `if (!sk_X509_value(xchain, i)) continue;` ?

**Review by andywolk — CHANGES_REQUESTED** [review]

File: src/stir_shaken_ssl.c (hunk 2 above)
Comment (mrtrev) [reply src/stir_shaken_ssl.c]: Done.  Looks like a copy/paste victim from elsewhere in the code (noticed lots of tab/space inconsistencies).

File: src/stir_shaken_ssl.c (hunk 5 above)
Comment (mrtrev) [reply src/stir_shaken_ssl.c:1816]: xc holds the certificate, this one is needed for the following function if the call was successful.

Comment (andywolk) [reply src/stir_shaken_ssl.c:1816]: Sure. I wanted to say `if (!(sc = sk_X509_value(xchain, i))) continue;` like the other ones. Looks clearer.

**Review by andywolk — CHANGES_REQUESTED** [review]

File: src/stir_shaken_ssl.c [hunk 6]
```diff
@@ -1995,6 +2018,60 @@ X509* stir_shaken_load_x509_from_file(stir_shaken_context_t *ss, const char *nam
...
+    }
+
+    while ((wcert = PEM_read_X509(fp, NULL, NULL, NULL))) {
+		if (!cert->x) {
+			cert->x = wcert;
+		}
+		else {
+			sk_X509_push(cert->xchain, wcert);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (mrtrev) [reply src/stir_shaken_ssl.c]: I've gone ahead and just reflowed the whole file so it no longer contains any tabs.  My editor would see them and assume we're using tabs for formatting.  This should help everyone going forward.

**Comment by andywolk** [comment]
@mrtrev Let's not create huge commits changing spacing in the whole codebase. Please remove it and fix what we touch only.

File: src/stir_shaken_ssl.c [hunk 7]
```diff
@@ -1792,12 +1806,20 @@ stir_shaken_status_t stir_shaken_x509_to_disk(stir_shaken_context_t *ss, X509 *x
...
-        if (i == 0) {
+        if (!PEM_write_X509(fp, x)) {
             snprintf(err_buf, sizeof(err_buf), "Error writing certificate to file %s", cert_full_name);
             stir_shaken_set_error(ss, err_buf, STIR_SHAKEN_ERROR_FILE_WRITE_2);
             goto fail;
         }
+
+        for (i = 0; i < sk_X509_num(xchain); i++) {
+            if (!(xc = sk_X509_value(xchain, i))) continue;
+            if (PEM_write_X509(fp, xc)) continue;
```
Comment (andywolk) [inline src/stir_shaken_ssl.c:1817]: Why continue? This looks unusual even if that's correct.
May be just
```
if (!PEM_write_X509(fp, xc)) {
            snprintf(err_buf, sizeof(err_buf), "Error writing certificate chain to file %s", cert_full_name);
            stir_shaken_set_error(ss, err_buf, STIR_SHAKEN_ERROR_FILE_WRITE_2);
            goto fail;
}
```

**Review by andywolk — CHANGES_REQUESTED** [review]

File: src/stir_shaken_ssl.c (hunk 7 above)
Comment (mrtrev) [reply src/stir_shaken_ssl.c]: done

**Review by andywolk — APPROVED** [review]

**Comment by andywolk** [comment]
Ok. Looks good to me.
@mrtrev Does it still work? And let's merge if so.

**Comment by mrtrev** [comment]
It still compiles.  I'll put it in prod and will advise tomorrow if nothing abnormal observed.

**Pull request merged**